odoo/fields.py

```diff
@@ -1077,10 +1077,10 @@ def __set__(self, records, value):
             # new records: no business logic
             new_records = records.browse(new_ids)
             with records.env.protecting(records._field_computed.get(self, [self]), records):
-                new_records.modified([self.name])
-                self.write(new_records, value)
                 if self.relational:
-                    new_records.modified([self.name])
+                    new_records.modified([self.name], before=True)
+                self.write(new_records, value)
+                new_records.modified([self.name])
```

Comment:
I think I understand the logic, but a little comment explaining why we have to call modified two times could be useful IMHO

Reply:
It's currently oral tradition :smile: 

It is best explained with a simple example: consider two sales order SO1 and SO2 with a bunch of lines.  The computed total amount on sales orders indirectly depends on the many2one field `order_id` linking lines to their sales order.  Now consider the following code:
```python
line = so1.line_ids[0]      # pick a line from SO1
line.order_id = so2         # move the line to SO2
```
In this situations, the total amount must be recomputed on *both* sales order: the line's order *before* the modification, and the line's order *after* the modification.  More generally, when modifying a relational field, you have to recompute what depends on the field's values before and after modification.  This is because the modification has an impact on the "data path" between a computed field and its dependency.  Note that this double call to `modified()` is only necessary for relational fields.

Reply:
Well, you can use the "oral" example as docstring then, sometimes it's clearer than a long technical explanation :smile: